src/dos/cdrom.cpp

```diff
@@ -268,3 +93,217 @@ void CDROM_Interface::LagDriveResponse() const
 
 	prev_ticks = PIC_Ticks;
 }
+
+void CDROM_Interface_Physical::CdReaderLoop()
+{
+	std::unique_lock<std::mutex> lock(mutex);
+	while (true) {
+		waiter.wait(lock, [&]{ return sectors_remaining > 0 || should_exit; });
+
+		if (should_exit) {
+			return;
+		}
+
+		if (!queue.IsRunning()) {
+			queue.Clear();
+			queue.Start();
+		}
+
+		// Copy these to local variables so that we can release the mutex lock.
+		const uint32_t local_sectors_remaining = sectors_remaining;
+		const uint32_t local_sector_start = current_sector;
+		lock.unlock();
+
+		// Slow call to read from the CDROM drive. Avoid holding any locks
+		const std::vector<int16_t> audio_samples = ReadAudio(local_sector_start, local_sectors_remaining);
+		constexpr size_t NumSamplesPerFrame = 2;
+		const size_t num_frames = audio_samples.size() / NumSamplesPerFrame;
+		const size_t sectors_read = audio_samples.size() / SAMPLES_PER_REDBOOK_FRAME;
+
+		// Convert from int16_t to float before enqueuing
+		std::vector<AudioFrame> audio_frames;
+		for (size_t i = 0; i < num_frames; ++i) {
+			const int16_t left = host_to_le16(audio_samples[i * NumSamplesPerFrame]);
+			const int16_t right = host_to_le16(audio_samples[(i * NumSamplesPerFrame) + 1]);
+			audio_frames.emplace_back(left, right);
+		}
+
+		queue.BulkEnqueue(audio_frames, audio_frames.size());
+
+		lock.lock();
+		if (mixer_channel && !is_paused && sectors_remaining > 0 && queue.IsRunning()) {
+			mixer_channel->Enable(true);
+		}
+
+		// Only modify these variables if they have not changed while we had the mutex lock released.
+		// Ex. The game did not want to switch tracks or stop the audio while we were reading/enqueuing.
+		if (sectors_remaining == local_sectors_remaining && current_sector == local_sector_start) {
+			sectors_remaining -= sectors_read;
+			current_sector += sectors_read;
+		}
+	}
+}
+
+void CDROM_Interface_Physical::InitAudio()
+{
+	if (mixer_channel) {
+		return;
+	}
+
+	// Each audio channel must have a unique name.
+	// Append a number after the CDAUDIO name.
+	// This way we don't conflict with CDROM_Interface_Image or multiple drives.
+	constexpr int MaxCdAudioChannel = 32;
```

Comment:
Ok, last one 😆 Consider moving this into `channel_names.h` and adding a comment above `CdAudio`, mentioning that we apply `_1`, `_2` etc. suffixes when we're dealing with multiple drives.

Reply:
Done